src/tools/wasm-merge.cpp

```diff
@@ -324,9 +324,26 @@ void renameInputItems(Module& input) {
 }
 
 void copyModuleContents(Module& input, Name inputName) {
-  // First, copy the regular module items (functions, globals) etc. which we
+  // Source map filename index mapping
+  std::vector<Index> indexMap;
+  std::unordered_map<std::string, Index> debugInfoFileIndices;
+  for (Index i = 0; i < merged.debugInfoFileNames.size(); i++) {
+    debugInfoFileIndices[merged.debugInfoFileNames[i]] = i;
+  }
+  for (Index i = 0; i < input.debugInfoFileNames.size(); i++) {
+    std::string file = input.debugInfoFileNames[i];
+    auto iter = debugInfoFileIndices.find(file);
+    if (iter == debugInfoFileIndices.end()) {
+      Index index = merged.debugInfoFileNames.size();
+      merged.debugInfoFileNames.push_back(file);
+      debugInfoFileIndices[file] = index;
+    }
+    indexMap.push_back(debugInfoFileIndices[file]);
+  }
+
+  // Copy the regular module items (functions, globals) etc. which we
   // have proper names for, and can just copy.
-  ModuleUtils::copyModuleItems(input, merged);
+  ModuleUtils::copyModuleItems(input, merged, &indexMap);
```

Comment:
I could be wrong, but the issue of mapping filename indices is potentially present in all calls of `copyFunction`, unless the entire module is cloned. That is, we can call `copyFunction` to copy a function from an arbitrary module to another. Given that, perhaps the above logic should be in `module-utils` so that all users use it? We may not have another user atm in the codebase, but doing it that way would avoid needing to refactor and maybe missing a bug if we do in the future.

Reply:
Oh, indeed!